src/saltan.8
.\"
.\" Copyright (c) 2021, 2025 Matthias Pressfreund
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: March 22 2025 $
.Dt SALTAN 8
.Os
.Sh NAME
.Nm saltan
.Nd
.Sy s Ns No shd Sy a Ns No uthentication Sy l Ns No og Sy t Ns No rack
.Sy a Ns No nd Sy n Ns No otify for
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8)
.Sh SYNOPSIS
.Nm
.Op Fl c Ar path
.Sh DESCRIPTION
.Nm
monitors the
.Xr sshd 8
authentication log file and instantly notifies
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8)
when it detects accepted and/or rejected authentication attempts.
To this end, it follows an extendable modular approach and executes a
configurable set of evaluation modules on each new entry in the authentication
log file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar path
Specifies the
.Ar path
to an alternative
.Nm
configuration file.
.El
.Sh CONFIGURATION
.Nm
reads its configuration either from the default configuration file
.Pa /etc/saltan/saltan.conf
or from the path specified with the
.Fl c
command line option.
.Pp
The following configuration options are accepted:
.Bl -tag -width Ds
.It Ic logfile Ar path
Specifies the
.Ar path
to the
.Xr sshd 8
authentication log file.
Defaults to
.Pa /var/log/authlog .
.It Ic pftbld Ar path
Specifies the
.Ar path
to the
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8)
binary.
Defaults to
.Pa /usr/local/sbin/pftbld .
.It Ic acceptsock | rejectsock Ar socket | Ic none
Specifies the
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8) Ns - Ns
related UNIX-domain
.Ar socket
to be notified in the event of an accepted or rejected authentication attempt.
Defaults to
.Ic none
which disables detection of corresponding authentication events.
At least one of the two options needs to be defined.
.It Ic acceptdir | rejectdir Ar path
Specifies the
.Ar path
to an evaluation module family's home directory.
Defaults to
.Pa /etc/saltan/accept
and
.Pa /etc/saltan/reject ,
respectively.
.It Ic acceptmods | rejectmods Ar name ...
Specifies a whitespace-separated list of module file names, located under
.Ic acceptdir
/
.Ic rejectdir ,
and executed in the order of appearance to detect authentication events.
Module names starting with a double underscore are ignored
.Pq e.g. Dv __dummy_module_1 .
Modules are activated only if the corresponding
.Ic acceptsock
/
.Ic rejectsock
is specified as well.
.El
.Sh MODULES
On startup,
.Nm
sources evaluation modules specified in the configuration file.
Each module consists of a short
.Xr ksh 1
script snippet that serves the following purposes:
.Bl -enum -offset indent -compact
.It
Determine whether a new entry in the
.Xr sshd 8
authentication log will be considered a relevant authentication event, and will
be treated further or should simply be ignored.
.It
Extract relevant information (e.g. the client's IP address) from the log
message.
.It
Build and return the notification message
.Nm
should send to
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8) .
.El
.Pp
.Nm
includes a set of ready-to-use modules providing base functionality which may
be adapted for individual requirements or used as templates for new log message
handlers.
.Pp
When modifying existing modules or writing new ones, the following aspects
should be considered:
.Bl -bullet -offset indent -compact
.It
As modules are sourced and executed inside the main
.Xr ksh 1 Ns - Ns
based script, they should not start with a shebang line
.Pq Dv #!... .
.It
.Nm
solely delivers the plain log message, i.e. that part of the log entry
following the sshd identifier indexed by the process ID
.Pq e.g. Dv sshd[69067] .
.It
A module must return 1 as soon as it turns out it is not responsible for
handling the given log message.
On the other hand, if it is responsible, it should end up returning 0 so that
.Nm
stops looking for a match.
.It
As modules are evaluated inside a function, new variables should be declared
.Dv local .
.It
The resulting
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html#MESSAGES notification
message must be provided by an
.Dv echo
command and has to always start with the client's IP address.
.It
If there is no
.Dv echo
command executed or the echoed message is empty,
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8)
will not be notified.
This is perfectly legal.
.It
Modules must be owned by
.Dv root
who is the only one granted write
.Pq Sq Dv w
permission.
At the same time, execute
.Pq Sq Dv x
permissions must not be granted to anyone at all.
.El
.Sh FILES
.Bl -tag -width "/usr/local/sbin/pftbld" -compact
.It Pa /etc/saltan/accept
The default
.Ic acceptdir .
.It Pa /etc/saltan/reject
The default
.Ic rejectdir .
.It Pa /etc/saltan/saltan.conf
The default
.Nm
configuration file.
.It Pa /usr/local/sbin/pftbld
The default path to the
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8)
binary.
.It Pa /var/log/authlog
The default
.Xr sshd 8
authentication log file.
.El
.Sh SEE ALSO
.Xr ksh 1 ,
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.8.html pftbld(8) ,
.Lk https://mpfr.net/man/pftbld/7.5-stable/pftbld.conf.5.html pftbld.conf(5) ,
.Xr sshd 8
.Sh HISTORY
.Nm
first appeared on
.Lk https://github.com/mpfr/saltan/ GitHub
in May, 2021.
.Sh AUTHORS
.An -nosplit
.Nm
was written by
.An Matthias Pressfreund .


src/saltan.sh
#!/bin/ksh
#
# Copyright (c) 2021, 2025 Matthias Pressfreund
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

usage()
{
	echo "usage: ${0##*/} [-c file]" >&2
	exit 1
}

err()
{
	echo "${0##*/}: $1" >&2
	exit 1
}

check_module()
{
	[[ -r $1 ]] || err "module '$1' does not exist"
	[[ $(stat -f '%Su' $1) = 'root' ]] || \
	    err "module '$1' is not owned by root"
	[[ $(stat -f '%Sp' $1) = '-'??-?'--'?'--' ]] || \
	    err "module '$1' does not have -??-?--?-- permissions"
}

notify_send()
{
	[[ -x ${PFTBLD} ]] || return
	echo -n "$2" | ${PFTBLD} -p $1 >/dev/null
}

run_module()
{
	local _mod=$1 _sock=$2 _msg
	shift 2
	_msg=$(eval "${_mod}" $*)
	if (($? == 0)); then
		[[ -n "${_msg}" ]] && notify_send ${_sock} "${_msg}"
		return 0
	fi
	return 1
}

while getopts c: arg; do
	case ${arg} in
	c)	CONFFILE=$OPTARG;;
	*)	usage;;
	esac
done

[[ -n ${CONFFILE} ]] || CONFFILE='/etc/saltan/saltan.conf'

while read line; do
	[[ ${line} = '#'* ]] && continue
	set -- ${line}
	(($# > 0)) || continue
	case $1 in
	'logfile')	LOGFILE=$2;;
	'pftbld')	PFTBLD=$2;;
	'acceptsock')	ACCEPTSOCK=$2;;
	'acceptdir')	ACCEPTDIR=$2;;
	'acceptmods')	shift; ACCEPTMODS=$*;;
	'rejectsock')	REJECTSOCK=$2;;
	'rejectdir')	REJECTDIR=$2;;
	'rejectmods')	shift; REJECTMODS=$*;;
	*)		err "invalid configuration parameter: $1"
	esac
done <${CONFFILE}

[[ -n ${LOGFILE} ]] || LOGFILE='/var/log/authlog'
[[ -r ${LOGFILE} ]] || err 'log file not found'

[[ -n ${PFTBLD} ]] || PFTBLD='/usr/local/sbin/pftbld'
[[ -x ${PFTBLD} ]] || err 'pftbld binary not found'

[[ -n ${REJECTSOCK} || -n ${ACCEPTSOCK} ]] || err 'no socket specified'

[[ -n ${ACCEPTDIR} ]] || ACCEPTDIR='/etc/saltan/accept'
[[ ${ACCEPTDIR} = 'none' ]] && ACCEPTDIR=''
[[ -d ${ACCEPTDIR} || -z ${ACCEPTDIR} ]] || \
    err 'accept modules directory not found'

[[ -n ${REJECTDIR} ]] || REJECTDIR='/etc/saltan/reject'
[[ ${REJECTDIR} = 'none' ]] && REJECTDIR=''
[[ -d ${REJECTDIR} || -z ${REJECTDIR} ]] || \
    err 'reject modules directory not found'

set -A modaccept
[[ -n ${ACCEPTDIR} ]] && for mod in ${ACCEPTMODS}; do
	[[ ${mod} = '__'* ]] && continue
	check_module ${ACCEPTDIR}/${mod}
	modaccept[((${#modaccept[@]}+1))]=$(cat ${ACCEPTDIR}/${mod})
done
set -A modreject
[[ -n ${REJECTDIR} ]] && for mod in ${REJECTMODS}; do
	[[ ${mod} = '__'* ]] && continue
	check_module ${REJECTDIR}/${mod}
	modreject[((${#modreject[@]}+1))]=$(cat ${REJECTDIR}/${mod})
done

tail -n0 -f ${LOGFILE} | while read -r line; do
	set -- ${line}
	shift 4
	[[ $1 = 'sshd-session['* || $1 = 'sshd-auth['* || $1 = 'sshd['* ]] || \
	    continue
	shift
	ret=1
	[[ -S ${REJECTSOCK} ]] && for mod in "${modreject[@]}"; do
		run_module "${mod}" ${REJECTSOCK} $*
		ret=$?
		((ret == 0)) && break
	done
	((ret == 0)) && continue
	[[ -S ${ACCEPTSOCK} ]] && for mod in "${modaccept[@]}"; do
		run_module "${mod}" ${ACCEPTSOCK} $*
		(($? == 0)) && break
	done
done
